airflow/plugins/operators/rectangle_realtime_data_operator.py

```diff
@@ -0,0 +1,214 @@
+
+    def cast_entities(self, entities):
+        # parse any information provided about fields that need to have dtype changed
+        # e.g. timestamps are strings, and latitude may be inferred as an int.
+        # note that due to a pandas bug, we first convert timestamps to
+        # a float, and then to an integer (hence the separate time_float_cast check)
+        # see: https://stackoverflow.com/a/60024263
+
+        return entities.astype(self.time_float_cast).astype(self.cast)
+
+    def get_entity_details(self, x):
+        """
+        Returns a list of dicts containing entity values to be added to a dataframe.
+        See https://gtfs.org/reference/realtime/v2/ for list of possible
+        entity values by message type.
+        """
+        entity = x.get("entity")
+        parsed_entity_details = []
+        if entity is not None:
+            for e in entity:
+                e_details = {
+                    col: self.pull_value(e, source)
+                    for col, source in self.entity_details.items()
+                }
+                parsed_entity_details.append(e_details)
+        return parsed_entity_details
+
+    def execute(self, context):
+        """Process a RT feed of the given type
+
+        Args:
+            rt_type (string): One of "alerts", "trip_updates", "vehicle_positions"
+            execution_date (date): The execution date being processed
+        """
+        fs = get_fs()
+
+        # get execution_date from context:
+        # https://stackoverflow.com/questions/59982700/airflow-how-can-i-access-execution-date-on-my-custom-operator
+        iso_date = str(context["execution_date"]).split("T")[0]
+
+        # fetch files ----
+        feed_files = self.fetch_bucket_file_names(iso_date)
+
+        # parse feeds ----
+        for feed, files in feed_files.items():
+            google_cloud_file_name = self.get_google_cloud_filename(feed, iso_date)
+            print("Creating " + google_cloud_file_name)
+            print("  parsing %s files" % len(files))
+
+            if len(files) > 0:
+                # fetch and parse RT files from bucket
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    fs.get(files, tmp_dir)
+                    all_files = [x for x in Path(tmp_dir).rglob("*") if not x.is_dir()]
+                    parsed_entities = [self.parse_pb(fn) for fn in all_files]
+
+                # convert protobuff objects to DataFrames
+                entities_dfs = [*map(self.rectangle_entities, parsed_entities)]
+                entities_dfs = [df for df in entities_dfs if df is not None]
+
+                print("  %s entities sub dataframes created" % len(entities_dfs))
+                if len(entities_dfs) > 0:
+                    entities_rectangle = pd.concat(entities_dfs)
+                    entities_rectangle.insert(0, "calitp_itp_id", int(feed[0]))
+                    entities_rectangle.insert(1, "calitp_url_number", int(feed[1]))
```

Comment:
Even with the changes I remember from our call (specifically using generators instead of lists), there's lots of in-memory operations to ultimately produce a dataframe that is written to parquet.

pandas uses either https://github.com/dask/fastparquet or https://github.com/apache/arrow/tree/master/python to write to parquet.

It seems like https://arrow.apache.org/docs/python/ipc.html#efficiently-writing-and-reading-arrow-data would be a good method for this. Something roughly like:

1.  For each feed, define the output stream and schema
2. Iterate over the files in the feed
3. Write each file in batches; if it's still too large, sub-divide into a sane number of records

Parallelization (across feeds) and asynchronous programming could help here.

Related question: is there a reason everything ends up in a single parquet file? Do you have an example of multiple file names within the same feed?

Reply:
Thanks Andrew, I do think we're going to need to move more in the streaming direction (and hoping we can work through that together) -- but I wanted to share more context, sorry in advance for length but I wanted to lay it all out in case it's helpful.

The terminology and the naming in the code here is a bit confusing. We're actually looping over an entire day's worth of files for each feed, so that's why there are multiple files. 

We start here:

**A**
[in`gtfs-data/rt`]
* [timestamp T on date D] 
   * [feed 1]
      * service_alerts protobuf file
      * trip_updates protobuf file
      * vehicle _positions protobuf file 
    * [feed 2]
      * service_alerts protobuf file
      * trip_updates protobuf file
      * vehicle _positions protobuf file 
 * [timestamp T+20s on date D] 
   * [feed 1]
      * service_alerts protobuf file
      * trip_updates protobuf file
      * vehicle _positions protobuf file 
    * [feed 2]
      * service_alerts protobuf file
      * trip_updates protobuf file
      * vehicle _positions protobuf file 
 
And we're converting to this:

**B**
[in `gtfs-data/rt-processed`]
* vehicle_positions 
   * [feed_1_vehicle_positions_date_D].parquet
   * [feed_2_vehicle_positions_date_D].parquet
* service_alerts
   * [feed_1_service_alerts_date_D].parquet
   * [feed_2_service_alerts_date_D].parquet 
* trip_updates
   * [feed_1_trip_updates_date_D].parquet
   * [feed_2_trip_updates_date_D].parquet 

So the current approach to the conversion to parquet is combining two goals:
1. Rectangling the data (converting from the JSON / key-value format to a nice rectangular dataframe)
2. Combining all 3k timestamp data points into just one file per day per feed

The raw protobuf files vary significantly in size and complexity from feed to feed. I think each input file itself is generally not too big, but when combined they start to get big in the output files.

Stepping back, I think that the **actual** goal from the pipeline perspective is just to get the data into BigQuery as an external table (I think that's the reason we need to convert file types, [BigQuery only supports certain file types in external tables from GCS](https://cloud.google.com/bigquery/external-data-cloud-storage)). 

I think that with BigQuery's [built-in JSON support](https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions) and [pyarrow's JSON support](https://arrow.apache.org/docs/python/json.html) it's possible we don't actually need to do the rectangling in Python here at all and can do all the UNNESTing stuff directly in BQ, but maybe it's possible to make the process more efficient via the approaches you mentioned. I definitely think parallelizing across feeds would help. 

I'm not sure how much we care about these intermediate parquet files, though I do think that Eric has used them for analysis recently and has built a significant amount of tooling to ingest them; we should notify him if we're going to change things substantially and move away from producing rectangled parquet files. 

A further piece of context: this pipeline is already running and working for the `vehicle_positions` data, so this current PR was basically just an attempt to apply that existing pipeline with minimal modifications to the other two data types. However, it seems that both in size and structure the `trip_updates` data in particular is probably not going to work. (I wanted to just to explain a bit why we're starting from this point; the vehicle positions data is smaller and already "more rectangular" than the other two data types, which might be why we started with it, but it also makes it harder to apply the same approach.)